### DISCUSSION TITLE
About the detail of trajectory of the railgun shell

### DISCUSSION THREAD

**Discussion by XiaotianMadurham** [FakeFishGames/Barotrauma#13865] [Code questions and modding support]
Hi, I am confused about the trajectory of the railgun shell. Is there any maximum speed limit on it once it is launched by the railgun? Additionally, does the shell experience any buoyancy? Also, what is the mass of the shell? I can only find a density value of 30 for a regular railgun shell. How does this relate to the mass?

Feel free to let me know if you need further adjustments or additional information. Below is a basic beyond visual range artillery system that I have created; however, due to my limited understanding of the correct trajectory function, it is currently only one-dimensional.
https://steamcommunity.com/sharedfiles/filedetails/?id=3044453304

**Comment by BlackMoons** [comment]
I dunno about mass, but 30 density means that it sinks quite rapidly (10 is neutral buoyancy)

I'll also note in the sub editor (And even some levels) its possible to fire a railgun straight up and be hit by it (same with coilguns) when it falls back down.

**Reply by XiaotianMadurham** [reply]
Thanks! my friend.

float volume = body.Mass / body.Density;
               
 float drag = speed * speed * WaterDragCoefficient * volume * Physics.NeutralDensity;
//very small drag on active projectiles to prevent affecting their trajectories much
if (body.FarseerBody.IsBullet) { drag *= 0.1f; }

if (moving || body.Density <= 10.0f)
            {
                Vector2 buoyancy = -GameMain.World.Gravity * forceFactor * volume * Physics.NeutralDensity;
                body.ApplyForce(buoyancy);
            }

The code snippet from Items.cs is sufficient for simulation. From experiments conducted in the game, I know that Physics.NeutralDensity is set to 10 (when shell density is 10, it floats and remains stationary in water). Gravity is set at 9.82, which I observed somewhere else in the code. The WaterDragCoefficient is approximately 50 for a shell, which seems excessively high (it might be more reasonable to set it at 5, since the shell may not be treat as a bullet when it simply sinks as an object).

Additionally, there is a maximum impulse of 60.8 in the game, indicating that even if a railgun delivers an impulse of 80, the effect is no different from 60.8. This is due to some speed limit code, but I have forgotten its exact location.

However, my current simulation is not very successful; there might be a bug.

**Reply by Vykn** [reply]
> same with coilguns

Coilgun bolts despawn after 3 seconds, so it's not possible to hit yourself with them by shooting straight up

**Comment by XiaotianMadurham** [comment]
[ballistic_test.zip](https://github.com/FakeFishGames/Barotrauma/files/15142430/ballistic_test.zip)
This is the current simulation model that i have, close but not accurate, Feel free to play around with it. maybe the problem is about the location where the drag be applied?
